DP-3195 Improve keyboard navigation of main menu

iansholtys opened the pull request:
This addresses a number of accessibility issues with the main menu:
* The buttons and links in the menu should all have `tabindex="-1"` except for the first one, which should be `0`
* Tab/Shift-Tab shouldn’t navigate through the menu items (that's what arrow keys are for, and it makes it take a very long time to skip past the menu)
* Menu bar and submenus should wrap around if you try to go past the end of a list
* Using the down or up arrows on a top-level menu item should expand that menu and move to the first or last items, respectively, rather than toggle the menu open and closed

With left/right and up/down arrows behaving so similarly, it was possible to consolidate and cut down on the code needed for those overall.

legostud commented on styleguide/source/assets/js/modules/mainNav.js:
```diff
@@ -32,64 +32,45 @@ export default function (window,document,$,undefined) {
+            focusIndexInDropdown = dropdownLinksLength;
           }
+          $dropdownLinks[focusIndexInDropdown-1].focus();
         } else {
```
needs a comment saying // Down arrow was used.  Otherwise, I naturally assumed any keyCode other than 38

legostud commented on styleguide/source/assets/js/modules/mainNav.js:
```diff
@@ -102,44 +83,29 @@ export default function (window,document,$,undefined) {
-        }
+        let leftArrow = e.keyCode === 37,
+            index = $topLevelLinks.index($topLevelLink),
+            linkCount = $topLevelLinks.length;
```
I prefer to declare my variables at the top of the section to make them easier to find.

legostud commented on styleguide/source/assets/js/modules/mainNav.js:
```diff
@@ -102,44 +83,29 @@ export default function (window,document,$,undefined) {
-      if(open || (typeof(e.keycode) !== "undefined" && e.keycode !== 9)) {
+      if(open || (typeof(e.keycode) !== "undefined")) {
         return;
       }
```
I think this is leftover code that doesn't do anything.  A blank return is usually done to block the code that follows from running.  There isn't any code following this.

legostud commented:
Also, I've mentioned this before, but I still feel that this code needs a spring cleaning, because it has been modified too many times by too many different people.  This could be a future Jira ticket.  

Ideally a few of us would sit down and record the desired behaviors to determine the proper JS class hooks needed and the proper events to listen too and then re-implement the code contained in a more organized and cleaner manner.

iansholtys commented on styleguide/source/_patterns/02-molecules/main-nav.twig:
```diff
@@ -5,23 +5,23 @@
           </div>
         {% else %}
-          <button id="{{ buttonId }}" class="ma__main-nav__top-link">{{ nav.text }}</button>
+          <button id="{{ buttonId }}" class="ma__main-nav__top-link" tabindex="{{ loop.first? 0 : -1 }}">{{ nav.text }}</button>
```
For this part I was mirroring the [W3C's Menubar example](https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html).

legostud replied:
I can't argue with a spec, but I still feel it's a poor UX to not be able to tab to each item in the main nav.

legostud commented:
I found a couple of issues that will need to be resolved.  

1. If you open the last menubar item, once you down arrow to the last item, the page starts scrolling instead of looping back to the top. Steps: focus on the first menubar item, press the right arrow four times followed by the up arrow that should then prevent the arrow keys from working in Chrome on Win10.

2. We should either address keyboard navigation on narrow screens or disable it.  Currently the left and right arrow move up and down and the up and down arrows slide the menu left and right.  Please note that the narrow screen can be seen by zooming the browser in so this could be the view on a desktop (ie: has a keyboard) someone with poor vision could be using.

iansholtys commented:
@legostud, I made a number of fixes and changes here:
* Removed legacy behavior for the last item in the last submenu which was causing it to get stuck open and not wrap to the top of the menu
* Adjusted breakpoint to 840px to match the current CSS value
* Perform keycode-to-action mapping at the start of the function so the rest of the code is easier to understand
* Consolidate similar uses of preventDefault()
* Adjust functionality so when the mobile menu is in use, keypresses have the following behaviors:
** Up/down navigates through the main nav, or if a subnav is open, through that. Focus wraps at the ends of the menus.
** Right arrow expands a submenu for the current menu item.
** Left arrow (or escape) exits out of the the current submenu.

I believe this should take care of the issues you identified.

clairesunstudio commented:
To make slider update value on drag, pass function to onUpdate prop
